docs/source/advanced.rst

```diff
@@ -229,11 +329,61 @@ type, and implements basic handling of the following cases out of the box:
 |                          | [.2, .9]]    |                |                |               |
 +--------------------------+--------------+----------------+----------------+---------------+
 
-If you find that your target is classified as ``"multiclass-multioutput"`` or ``"unknown"``, you will have to
-implement your own data processing routine.
+The supported cases are handled by the default implementation of ``target_encoder``.
+The default implementations are available for use as :py:class:`scikeras.utils.transformers.ClassifierLabelEncoder`
+and :py:class:`scikeras.utils.transformers.RegressorTargetEncoder` for
+:py:class:`scikeras.wrappers.KerasClassifier` and :py:class:`scikeras.wrappers.KerasRegressor` respectively.
 
-For a complete examples implementing custom data processing, see the examples in the
-:ref:`tutorials` section.
+As per the table above, if you find that your target is classified as
+``"multiclass-multioutput"`` or ``"unknown"``, you will have to implement your own data processing routine.
+
+Whole dataset manipulation via data_transformer
++++++++++++++++++++++++++++++++++++++++++++++++
+
+This is the last step before passing the data to Keras, and it allows for the greatest
+degree of customization because SciKeras does not make any assumptions about the output data
+and passes it directly to :py:func:`tensorflow.keras.Model.fit`.
+Its signature is ``dataset_transformer.fit_transform((X, y, sample_weight))``,
```

Comment:
It's signature is this:

``` python
from sklearn.base import BaseEstimator

class DatasetTransformer(BaseEstimator):
    def fit(self, X, y, sample_weight=None) -> "DatasetTransformer":
        ...
        return self

    def transform(self, X, y, sample_weight):  # return a valid input for keras.Model.fit
        ...
        return X, y, sample_weight  # option 1
        return tensorflow_dataset  # option 2
```

Reply:
I may have design it wrong, and I'm not a fan of the tuple input either, but I believe it is required so that these can be chained in a pipeline.

With the signature you are proposing, this fails:

```python3
from sklearn.base import BaseEstimator
from sklearn.pipeline import make_pipeline

class DatasetTransformer(BaseEstimator):
    def fit(self, X, y, sample_weight=None) -> "DatasetTransformer":
        return self

    def transform(self, X, y, sample_weight):  # return a valid input for keras.Model.fit
        return X, y, sample_weight

p = make_pipeline(DatasetTransformer(), DatasetTransformer())

X = [1]
y = [1]

p.fit_transform(X, y)
```

If instead you accept a tuple as your input, then you can chain them:

```python3
class DatasetTransformer(BaseEstimator):
    def fit(self, X, y, sample_weight=None) -> "DatasetTransformer":
        return self

    def transform(self, data):  # return a valid input for keras.Model.fit
        X, y, sample_weight = data
        return X, y, sample_weight

p = make_pipeline(DatasetTransformer(), DatasetTransformer())

X = [1]
y = [1]

p.fit_transform((X, y, None))
```

Am I missing something?

Reply:
With the example, I only meant "illustrate the signature with code, not with words." I am unsure what's happening with tuple inputs.

Reply:
Oh I see. Yes, that is a good suggestion.

Reply:
Maybe it'd be clearer to have the user return a dict. I'd also modify the tuple input to this transform; why not pass keyword arguments?

``` python
ValidKerasInput = Union[np.ndarray, tf.Dataset, ...]

def transform(
    self,
    X: ValidKerasInput,
    y: ValidKerasInput,
    sample_weight: Optional[ValidKerasInput]=None
) -> dict[str, ValidKerasInput]:
    ...
    return {"X": X, "y": y, "sample_weight": None}
```

Reply:
That signature would work (and is ideal) for a single transformer. I think the issue is that sklearn assumes the signature `transform(X)` (i.e. a single argument), so to be compatible with the input must be a single argument. And to chain, the input and output must be identical.

That said, we could swap from a tuple to a dict. As long as it passes this test it should work:

```python3
p = make_pipeline(DatasetTransformer(), DatasetTransformer())
p.fit_transform((X, y, sample_weight)) # or {"X": X, "y"....}
```

Reply:
> I think the issue is that sklearn assumes the signature transform(X) (i.e. a single argument), so to be compatible with the input must be a single argument. And to chain, the input and output must be identical.

Yes, that's the formal Scikit-learn API specification. But, that specification is not enforced. I wouldn't complain if you broke that specification: I can't imagine users will chain `DatasetTransformer`s together, and I think multiple inputs will enhance usability.

Also, conforming to API specifications is a good practice, so I won't complain if you follow the API specification.

Reply:
It's not just about the API specification, in my very limited testing I already found use cases for chaining, like [this part of the notebooks](https://www.adriangb.com/scikeras/refs/pull/167/merge/notebooks/DataTransformers.html#5.-Ragged-datasets-with-tf.data.Dataset):


```python3
class RaggedClassifier(KerasClassifier):

    @property
    def dataset_transformer(self):
        t1 = FunctionTransformer(ragged_transformer)
        t2 = super().dataset_transformer  # ClassWeightDataTransformer
        t3 = FunctionTransformer(dataset_transformer)
        return make_pipeline(t1, t2, t3)
```

Reply:
I was able to remove `dummy`/the second parameter by adding a `"passthrough"` step in the pipeline. This is still a quirk, but a lesser evil (I think).

Reply:
I'm not sure I understand. It took a while to understand the `passthrough` parameter (I think I understand it). I can understand a tuple/dict right away, but the "passthrough" option requires some thought and digging.

I'm a fan of passing a dict around:

``` python
p = make_pipeline(DatasetTransformer(), DatasetTransformer())
data = {"X": X, "y": y, "sample_weight": sample_weight}
p.fit(data)
p.transform(data)
p.fit_transform(data)
```

I slightly prefer passing a dict around because the keys aren't fixed, and a tuple has a static length.

Reply:
A dict sounds good, thank you for the suggestion. As a side note, I dug into the issue a bit more, it seems that being able to modify only X (vs. y and sample_weight like we do here) is a [known design flaw](https://github.com/scikit-learn/scikit-learn/issues/3855#issuecomment-63213164) in sklearn. This makes me feel a bit more comfortable "adapting" the API in this fashion.

Regarding the `"passthrough"`/`dummy` thing: from the [Pipeline docs](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html), the final estimator _must_ implement `fit`. And `fit` is expected to have the signature `fit(X, y)` and is called like that ([here](https://github.com/scikit-learn/scikit-learn/blob/b3ea3ed6a09fe774dfc5160a65172b1bacbb2a82/sklearn/pipeline.py#L346)). But since we are including `y` in `data`, it does not make sense to have the signature `fit(data, y=None)` and then override `y` in the body of the function. Hence why I originally renamed the parameter from `y` to `dummy` to avoid confusion. The alternative, which is a fix specifically for pipelines but may leave room for issues elsewhere, is to use the better signature of `fit(data)` but force the last estimator in the pipeline to be `"passthrough"`/`None`. This means that all of these transformers are "known" to be transformers and so they are called with just `X` (here).

I'm a bit torn here, on the one hand, I don't like having a parameter called `dummy`. On the other hand, I think the `"passthrough"` solution is more fragile (and it isn't very nice either).

Reply:
> ` fit(data, y=None)`

Why not `fit(data, **ignored)` or `fit(data, **empty)`? If you want, you can even write `assert len(empty) == 0`.

Reply:
I agree. Let me try implementing this tomorrow.

For the dict, do you envision passing/returning `None` or just not including the key?

Also, a minor nit, but sklearn uses `X` (capital) while Keras uses `x` (lower). There's something to be said for using `x`, since it allows for `self.model_.fit(**data, ...)`. Do you have a preference for `X` or `x`?

Reply:
I think it's easiest to say "the keys in the returned dictionary should be valid keywords to keras.Model.fit".

Let's take a transformer that does nothing and returns the input, in-modified.

Then, `x` should be preferred because it's a valid keyword argument to keras.Model.fit. I suspect that answers the no key question.

Reply:
If we are going to go with dicts, what are your thoughts on passing all `Model.fit`/`Model.predict` `**kwargs`?

The pro is that it would allow nearly arbitrary user logic injection. For example, #131 could be resolved by creating a transformer that siphons off validation data from `x` & `y` into `validation_dataset`. This would be done after applying target encoding, etc. so it should be compatible with all of sklearn and Keras.

The con is that I think the interface could be messy.

Reply:
How would it be messy? Arbitrary user logic sounds like a good feature for something basic like this.

> If we are going to go with dicts, what are your thoughts on passing all Model.fit/Model.predict **kwargs?

Doesn't a user-passed kwarg already passed to `fit`? I'm thinking of `scikeras_model.fit(...,  batch_size=42)`. That's in `kwargs` and gets passed to `model_.fit`, right?

Reply:
> How would it be messy?

It's minor, but it would now be a dict with dozens of keys, some values being arrays, etc. It can easily be documented, but I can see how at first glance it might be more confusing than a 3 element tuple or a 3 key dict.

> Doesn't a user-passed kwarg already passed to fit?

Yes, but these parameters can't be calculated based on the data, which becomes especially problematic if doing cross-valdation and such. This is the basic problem in #131.

Of course, if users want to pass a fixed parameter like `batch_size=42` they can do so via `fit` (or better yet, via the constructor).

Reply:
> It's minor, but it would now be a dict with dozens of keys,

Why would it have to be dozens of keys? I'm thinking of this code:

``` python
# line 885 of wrappers.py
transform_kwargs = self.dataset_transformer_.transform(x=X, y=y, sample_weight=sample_weight)
kwargs.update(transform_kwargs)
self._fit_keras_model(**kwargs)
```

Reply:
I think that implementation in https://github.com/adriangb/scikeras/pull/167#discussion_r576512695 is not compatible with the sklearn ecosystem, including pipelines, which I think are in important feature. This would be fine:

```python3
self.dataset_transformer_.transform(dict(x=X, y=y, sample_weight=sample_weight))
```

Regarding the multiple keys, what I was thinking was to pass _all_ `Model.fit` keys. In `_fit_keras_model`:

https://github.com/adriangb/scikeras/blob/6eee3c4a19b28694758e02aab27af9222ff82db8/scikeras/wrappers.py#L480-L494

This means that users can not only add keys, but also modify existing keys, to their liking.

Regarding the `dozens of keys` comment, I wasn't referring to the keys assigned via `kwargs.update(transform_kwargs)` above, I was more thinking that when a user creates a transformer like:

```python
def transform(data: Dict[str, Any) -> Dict[str, Any]:
    len(data.keys())  # many keys, array-values, etc.
    ...

# for context:
class SomeCustomEstimator(BaseWrapper):
    @property
    def data_transformer(self):
        return FunctionTransformer(transformer)
```

Reply:
> This means that users can not only add keys, but also modify existing keys, to their liking.

Yes, I think that's a good idea. It also leads to simple documentation (something like "the dataset transformer is passed a dict. Normally, this dictionary is unmodified and passed directly to keras.Model.fit. If desired, it's possible to modify this dictionary through a Scikit-learn transformation: [example]."

Reply:
Great. I'm going to loop back to #131 and #159 to see if the issue reporters have any input on this implementation.